Makefile

```diff
@@ -194,32 +193,64 @@ $(TERRAFORM):
-	@$(OK) generating provider schema for $(TERRAFORM_PROVIDER_SOURCE) $(TERRAFORM_PROVIDER_VERSION)
+	@echo 'data "aws_partition" "example" {}' > $(TERRAFORM_WORKDIR)/main.tf
+	@$(TERRAFORM) -chdir=$(TERRAFORM_WORKDIR) init -plugin-dir $(ROOT_DIR)/upstream/terraform-plugin-dir > $(TERRAFORM_WORKDIR)/terraform-logs.txt 2>&1
+	@$(TERRAFORM) -chdir=$(TERRAFORM_WORKDIR) providers schema -json=true > $(TERRAFORM_PROVIDER_SCHEMA) 2>> $(TERRAFORM_WORKDIR)/terraform-logs.txt
+	@$(OK) generated provider schema for $(TERRAFORM_PROVIDER_SOURCE) from the upstream git submodule
 
-pull-docs:
-	rm -fR "$(WORK_DIR)/$(notdir $(TERRAFORM_PROVIDER_REPO))"
-	git clone -c advice.detachedHead=false --depth 1 --filter=blob:none --branch "v$(TERRAFORM_PROVIDER_VERSION)" --sparse "$(TERRAFORM_PROVIDER_REPO)" "$(WORK_DIR)/$(notdir $(TERRAFORM_PROVIDER_REPO))";
-	@git -C "$(WORK_DIR)/$(notdir $(TERRAFORM_PROVIDER_REPO))" sparse-checkout set "$(TERRAFORM_DOCS_PATH)"
+# Add sentinel files for tracking which targets need to be rebuilt
+UPSTREAM := $(ROOT_DIR)/.upstream.sentinel
 
-generate.init: $(TERRAFORM_PROVIDER_SCHEMA) pull-docs
+# Applies all the patches to the upstream files, but does not commit them.
+# Rebuild if the patches change, or upstream moves to a new HEAD, with some extra logic in the shell script
+$(UPSTREAM): $(wildcard patches/*) $(shell ./scripts/upstream.sh file_target)
+	@$(INFO) Patching upstream terraform provider source
+	@$(ROOT_DIR)/scripts/upstream.sh init -f
+	@touch $@
+	@$(OK) Patched upstream terraform provider source
+
+# Apply patches to upstream tf provider source
+upstream: $(UPSTREAM)
+
+# Alias to build the upstream terraform provider
+upstream.build: $(TERRAFORM_PROVIDER)
+
+# Overwrite the check-diff target from the build submodule to exclude the upstream submodule.
+# TODO(mbbush): merge this change into the build submodule in a backwards-compatible way once the use cases are clear.
+check-diff: generate
+	@$(INFO) checking that branch is clean
+	@if git status --porcelain --ignore-submodule=all | grep . ; then $(ERR) There are uncommitted changes after running make generate. Please ensure you commit all generated files in this branch after running make generate. && false; else $(OK) branch is clean; fi
+	@$(INFO) checking that the build submodule is clean
+	@if git status --porcelain build | grep . ; then $(ERR) There are uncommitted changes in the build submodule. Please ensure the build submodule is checked out at the desired commit. && false; else $(OK) build submodule is clean; fi
+	@$(INFO) checking that the upstream submodule
+	@if git status --porcelain --ignore-submodule=dirty | grep . ; then $(ERR) There are new commits in the upstream submodule. Please ensure all changes are written to patch files using scripts/upstream.sh. && false; else $(OK) upstream submodule is clean; fi
+
+
+.PHONY: upstream upstream.build
+
+generate.init: $(UPSTREAM) $(TERRAFORM_PROVIDER) $(TERRAFORM_PROVIDER_SCHEMA)
+
+build.init: $(UPSTREAM)
```

Comment:
I think we are considering this will affect all build phases/targets of the provider?

Reply:
Yes, my goal was to have the upstream target run before any of the other build targets. If you know of a better way to do that than specifying them all, I'd love to hear it.